---
title: "Smoothclust Tutorial"
author: 
  - name: Lukas M. Weber
    affiliation: "Boston University"
package: smoothclust
output: 
  BiocStyle::html_document: 
    toc_float: true
vignette: >
  %\VignetteIndexEntry{Smoothclust Tutorial}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


# Introduction

`smoothclust` is a spatial clustering algorithm for spatial transcriptomics data based on the principle of smoothing expression measurements across neighboring spatial locations. The algorithm can be used to define spatial domains consisting of a consistent mixture of cell types with smooth spatial boundaries.


# Installation

Install R package from GitHub: `remotes::install_github("lmweber/smoothclust")`


# Input data format

Input data can be provided either as a `SpatialExperiment` object or a numeric matrix containing expression values. See help file (`?smoothclust`) for details. In the example workflow below, we assume the input is in `SpatialExperiment` format.


# Tutorial

Example demonstrating how to run `smoothclust` on a 10x Genomics Visium dataset (human DLPFC dataset).

```{r, message=FALSE}
library(smoothclust)
library(STexampleData)
library(scuttle)
library(scran)
library(scater)
library(ggspavis)
```


Load dataset from `STexampleData` package:

```{r}
# load data
spe <- Visium_humanDLPFC()

# keep spots over tissue
spe <- spe[, colData(spe)$in_tissue == 1]

dim(spe)
```


Run `smoothclust` with default parameter settings:

```{r, results="hide"}
# run smoothclust
spe <- smoothclust(spe)
```


Calculate log-transformed normalized counts (logcounts) on smoothed values:

```{r}
# calculate logcounts
spe <- logNormCounts(spe)
```


Run clustering:

```{r}
# preprocessing steps for clustering

# remove mitochondrial genes
is_mito <- grepl("(^MT-)|(^mt-)", rowData(spe)$gene_name)
table(is_mito)
spe <- spe[!is_mito, ]
dim(spe)

# select top highly variable genes (HVGs)
dec <- modelGeneVar(spe)
top_hvgs <- getTopHVGs(dec, prop = 0.1)
length(top_hvgs)
spe <- spe[top_hvgs, ]
dim(spe)
```

```{r}
# dimensionality reduction

# compute PCA on top HVGs
set.seed(123)
spe <- runPCA(spe)
```

```{r}
# run k-means clustering
set.seed(123)
k <- 5
clust <- kmeans(reducedDim(spe, "PCA"), centers = k)$cluster
table(clust)
colLabels(spe) <- factor(clust)
```


Plot clusters / spatial domains:

```{r}
pal8 <- "libd_layer_colors"
pal36 <- unname(palette.colors(36, "Polychrome 36"))

# plot clusters
plotSpots(spe, annotate = "label", palette = pal8)

# plot reference cluster labels
plotSpots(spe, annotate = "ground_truth", palette = pal8)
```


# Session information

```{r}
sessionInfo()
```
